File: clients/java-sdk/src/main/java/com/example/iggy/Schemas.java
package com.example.iggy;

import com.google.protobuf.DescriptorProtos;
import com.google.protobuf.Descriptors;
import com.google.protobuf.DynamicMessage;
import org.apache.avro.Schema;
import org.apache.avro.generic.GenericData;
import org.apache.avro.generic.GenericDatumReader;
import org.apache.avro.generic.GenericDatumWriter;
import org.apache.avro.generic.GenericRecord;
import org.apache.avro.io.BinaryDecoder;
import org.apache.avro.io.BinaryEncoder;
import org.apache.avro.io.DecoderFactory;
import org.apache.avro.io.EncoderFactory;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;

/**
 * Shared Avro and Protobuf schema helpers for the Iggy format examples.
 *
 * <p>The Avro schema mirrors {@code resources/schemas/event.avsc} and the Protobuf
 * descriptor mirrors {@code resources/schemas/event.proto}.  Both are defined here
 * so that the producer/consumer example classes and their tests can import from a
 * single place without requiring a code-generation step.
 */
public final class Schemas {

    private Schemas() {}

    // -----------------------------------------------------------------------
    // Avro
    // -----------------------------------------------------------------------

    /** Avro schema string that mirrors {@code schemas/event.avsc}. */
    public static final String EVENT_AVRO_SCHEMA_JSON =
            "{"
            + "\"type\":\"record\","
            + "\"name\":\"Event\","
            + "\"namespace\":\"iggy.example\","
            + "\"fields\":["
            + "  {\"name\":\"id\",  \"type\":\"int\"},"
            + "  {\"name\":\"text\",\"type\":\"string\"},"
            + "  {\"name\":\"ts\",  \"type\":\"string\"}"
            + "]}";

    public static final Schema AVRO_SCHEMA =
            new Schema.Parser().parse(EVENT_AVRO_SCHEMA_JSON);

    /**
     * Serialize an event to schemaless Avro bytes.
     *
     * @param id   message id
     * @param text human-readable body
     * @param ts   ISO-8601 UTC timestamp
     * @return binary Avro encoding of the event
     */
    public static byte[] avroSerialize(int id, String text, String ts) throws IOException {
        GenericRecord record = new GenericData.Record(AVRO_SCHEMA);
        record.put("id", id);
        record.put("text", text);
        record.put("ts", ts);

        ByteArrayOutputStream out = new ByteArrayOutputStream();
        BinaryEncoder encoder = EncoderFactory.get().binaryEncoder(out, null);
        new GenericDatumWriter<GenericRecord>(AVRO_SCHEMA).write(record, encoder);
        encoder.flush();
        return out.toByteArray();
    }

    /**
     * Deserialize schemaless Avro bytes back to a {@link GenericRecord}.
     *
     * @param data binary Avro bytes produced by {@link #avroSerialize}
     * @return the decoded record
     */
    public static GenericRecord avroDeserialize(byte[] data) throws IOException {
        BinaryDecoder decoder =
                DecoderFactory.get().binaryDecoder(new ByteArrayInputStream(data), null);
        return new GenericDatumReader<GenericRecord>(AVRO_SCHEMA).read(null, decoder);
    }

    // -----------------------------------------------------------------------
    // Protobuf
    // -----------------------------------------------------------------------

    private static final Descriptors.Descriptor EVENT_DESCRIPTOR;

    static {
        try {
            DescriptorProtos.FileDescriptorProto fileProto =
                    DescriptorProtos.FileDescriptorProto.newBuilder()
                            .setName("iggy_example_event.proto")
                            .setPackage("iggy.example")
                            .setSyntax("proto3")
                            .addMessageType(
                                    DescriptorProtos.DescriptorProto.newBuilder()
                                            .setName("Event")
                                            .addField(field("id",   1, DescriptorProtos.FieldDescriptorProto.Type.TYPE_INT32))
                                            .addField(field("text", 2, DescriptorProtos.FieldDescriptorProto.Type.TYPE_STRING))
                                            .addField(field("ts",   3, DescriptorProtos.FieldDescriptorProto.Type.TYPE_STRING))
                                            .build())
                            .build();
            Descriptors.FileDescriptor fileDescriptor =
                    Descriptors.FileDescriptor.buildFrom(
                            fileProto, new Descriptors.FileDescriptor[0]);
            EVENT_DESCRIPTOR = fileDescriptor.findMessageTypeByName("Event");
        } catch (Descriptors.DescriptorValidationException e) {
            throw new ExceptionInInitializerError(e);
        }
    }

    private static DescriptorProtos.FieldDescriptorProto field(
            String name, int number, DescriptorProtos.FieldDescriptorProto.Type type) {
        return DescriptorProtos.FieldDescriptorProto.newBuilder()
                .setName(name)
                .setNumber(number)
                .setType(type)
                .setLabel(DescriptorProtos.FieldDescriptorProto.Label.LABEL_OPTIONAL)
                .build();
    }

    /** Returns the descriptor for the {@code iggy.example.Event} Protobuf message. */
    public static Descriptors.Descriptor eventDescriptor() {
        return EVENT_DESCRIPTOR;
    }

    /**
     * Serialize an event to Protobuf wire-format bytes.
     *
     * @param id   message id
     * @param text human-readable body
     * @param ts   ISO-8601 UTC timestamp
     * @return Protobuf binary encoding of the event
     */
    public static byte[] protobufSerialize(int id, String text, String ts) {
        return DynamicMessage.newBuilder(EVENT_DESCRIPTOR)
                .setField(EVENT_DESCRIPTOR.findFieldByName("id"),   id)
                .setField(EVENT_DESCRIPTOR.findFieldByName("text"), text)
                .setField(EVENT_DESCRIPTOR.findFieldByName("ts"),   ts)
                .build()
                .toByteArray();
    }

    /**
     * Deserialize Protobuf bytes back to a {@link DynamicMessage}.
     *
     * @param data Protobuf wire-format bytes produced by {@link #protobufSerialize}
     * @return the decoded message
     */
    public static DynamicMessage protobufDeserialize(byte[] data)
            throws com.google.protobuf.InvalidProtocolBufferException {
        return DynamicMessage.parseFrom(EVENT_DESCRIPTOR, data);
    }
}

File: clients/rust/src/schemas.rs
//! Shared Avro and Protobuf schema helpers for the Iggy format examples.
//!
//! The Avro schema mirrors `schemas/event.avsc` and the Protobuf struct
//! mirrors `schemas/event.proto`.  Both are defined here so that the
//! producer/consumer example binaries and their `#[cfg(test)]` tests can
//! import from a single place without requiring a code-generation step.

use apache_avro::{from_avro_datum, to_avro_datum, Schema};
use apache_avro::types::Value;

// ---------------------------------------------------------------------------
// Avro
// ---------------------------------------------------------------------------

/// Raw Avro schema JSON string mirroring `schemas/event.avsc`.
pub const EVENT_AVRO_SCHEMA_JSON: &str = r#"
{
  "type":      "record",
  "name":      "Event",
  "namespace": "iggy.example",
  "fields": [
    {"name": "id",   "type": "int"},
    {"name": "text", "type": "string"},
    {"name": "ts",   "type": "string"}
  ]
}
"#;

/// Returns the parsed [`Schema`] for the `Event` Avro record.
pub fn avro_schema() -> Schema {
    Schema::parse_str(EVENT_AVRO_SCHEMA_JSON).expect("EVENT_AVRO_SCHEMA_JSON is always valid")
}

/// Serialize an event to schemaless Avro bytes.
pub fn avro_serialize(id: i32, text: &str, ts: &str) -> Result<Vec<u8>, apache_avro::Error> {
    let schema = avro_schema();
    let value = Value::Record(vec![
        ("id".to_string(),   Value::Int(id)),
        ("text".to_string(), Value::String(text.to_string())),
        ("ts".to_string(),   Value::String(ts.to_string())),
    ]);
    to_avro_datum(&schema, value)
}

/// Deserialize schemaless Avro bytes back to a `(id, text, ts)` tuple.
pub fn avro_deserialize(data: &[u8]) -> Result<(i32, String, String), apache_avro::Error> {
    let schema = avro_schema();
    let mut cursor = std::io::Cursor::new(data);
    let value = from_avro_datum(&schema, &mut cursor, None)?;
    match value {
        Value::Record(fields) => {
            let mut id: i32 = 0;
            let mut text = String::new();
            let mut ts = String::new();
            for (name, val) in fields {
                match (name.as_str(), val) {
                    ("id",   Value::Int(v))    => id   = v,
                    ("text", Value::String(v)) => text = v,
                    ("ts",   Value::String(v)) => ts   = v,
                    _ => {}
                }
            }
            Ok((id, text, ts))
        }
        other => Err(apache_avro::Error::new(apache_avro::error::Details::DeserializeValue(
            format!("expected Record, got {:?}", other),
        ))),
    }
}

// ---------------------------------------------------------------------------
// Protobuf
// ---------------------------------------------------------------------------

/// `Event` Protobuf message struct mirroring `schemas/event.proto`.
///
/// The `prost::Message` derive macro generates efficient encode/decode
/// implementations without requiring a `protoc` code-generation step.
#[derive(Clone, PartialEq, prost::Message)]
pub struct Event {
    /// Monotonically incrementing message ID.
    #[prost(int32, tag = "1")]
    pub id: i32,
    /// Human-readable message body.
    #[prost(string, tag = "2")]
    pub text: String,
    /// ISO-8601 UTC timestamp of the event.
    #[prost(string, tag = "3")]
    pub ts: String,
}

/// Serialize an event to Protobuf wire-format bytes.
pub fn protobuf_serialize(id: i32, text: &str, ts: &str) -> Vec<u8> {
    use prost::Message;
    Event { id, text: text.to_string(), ts: ts.to_string() }.encode_to_vec()
}

/// Deserialize Protobuf bytes back to an [`Event`].
pub fn protobuf_deserialize(data: &[u8]) -> Result<Event, prost::DecodeError> {
    use prost::Message;
    Event::decode(data)
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    // Avro tests -----------------------------------------------------------

    #[test]
    fn avro_roundtrip_produces_original_fields() {
        let data = avro_serialize(1, "hello from Rust Avro producer", "2024-01-01T00:00:00Z")
            .expect("serialize");
        let (id, text, ts) = avro_deserialize(&data).expect("deserialize");
        assert_eq!(id, 1);
        assert_eq!(text, "hello from Rust Avro producer");
        assert_eq!(ts, "2024-01-01T00:00:00Z");
    }

    #[test]
    fn avro_serialized_bytes_are_not_empty() {
        let data = avro_serialize(42, "test", "2024-01-01T00:00:00Z").expect("serialize");
        assert!(!data.is_empty());
    }

    #[test]
    fn avro_different_records_produce_different_bytes() {
        let a = avro_serialize(1, "first",  "2024-01-01T00:00:00Z").expect("serialize");
        let b = avro_serialize(2, "second", "2024-01-01T00:00:00Z").expect("serialize");
        assert_ne!(a, b);
    }

    #[test]
    fn avro_successive_records_carry_incrementing_ids() {
        for i in 1..=3 {
            let data = avro_serialize(i, "msg", "2024-01-01T00:00:00Z").expect("serialize");
            let (id, _, _) = avro_deserialize(&data).expect("deserialize");
            assert_eq!(id, i);
        }
    }

    // Protobuf tests -------------------------------------------------------

    #[test]
    fn protobuf_roundtrip_restores_all_fields() {
        let data = protobuf_serialize(1, "hello from Rust Protobuf producer", "2024-01-01T00:00:00Z");
        let event = protobuf_deserialize(&data).expect("deserialize");
        assert_eq!(event.id, 1);
        assert_eq!(event.text, "hello from Rust Protobuf producer");
        assert_eq!(event.ts, "2024-01-01T00:00:00Z");
    }

    #[test]
    fn protobuf_serialized_bytes_are_not_empty() {
        let data = protobuf_serialize(42, "test", "2024-01-01T00:00:00Z");
        assert!(!data.is_empty());
    }

    #[test]
    fn protobuf_different_ids_produce_different_bytes() {
        let a = protobuf_serialize(1, "same text", "2024-01-01T00:00:00Z");
        let b = protobuf_serialize(2, "same text", "2024-01-01T00:00:00Z");
        assert_ne!(a, b);
    }

    #[test]
    fn protobuf_successive_events_carry_incrementing_ids() {
        for i in 1..=3i32 {
            let data = protobuf_serialize(i, "msg", "2024-01-01T00:00:00Z");
            let event = protobuf_deserialize(&data).expect("deserialize");
            assert_eq!(event.id, i);
        }
    }

    #[test]
    fn protobuf_default_values_roundtrip() {
        let data = protobuf_serialize(0, "", "");
        let event = protobuf_deserialize(&data).expect("deserialize");
        assert_eq!(event.id, 0);
        assert_eq!(event.text, "");
        assert_eq!(event.ts, "");
    }
}
